### PR TITLE
Paypal sub button creates an alert when there are errors fix #6401

### LINKED ISSUE #6401: immediate cancellation of new PayPal subscription shows the user a raw error message
If you create a subscription with PayPal and then try to cancel it immediately, the notification that you can't do that yet appears as a raw error message in a page by itself, instead of in a dialog screen with an "ok" button to dismiss it.

```
{"err":"\"The plan hasn't activated yet (due to a PayPal bug). It will begin 2015-12-26T10:00:00Z, after which you can cancel to retain your full benefits\""}
```

screen shot 2015-12-26 at 12 20 44 pm

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Should fix #6401

The main body of the fix just handles any error messages coming back from Paypal.  I opted to pull out error_description, it looks to be the most informative message, and pop that into an alert.

In addition I changed a bit of paypalBillingSetup since that didn't appear to be working properly or give you the keys you'd want for testing.

Hope this is okay!

### CODE REVIEW

**Comment by SabreCat** [comment]
Taking a look at this, the functions added and modified could use test coverage. Let us know if you need a hand navigating the test framework!

**Comment by ojeytonwilliams** [comment]
@SabreCat I might need a little help, actually!

Am I right in thinking that I should be writing two tests - one karma test for paymentServices.js and a mocha test for paypal.js?  The impression I got from looking through the tests was that the karma test should be similar to the tests in test/spec/services and the mocha test should be something like the tests in test/server_side/controllers

If that's broadly the right idea, I'll see what I can do from here.

**Comment by SabreCat** [comment]
You've got it! Karma (test/spec/*) is for our Angular tests, and Mocha (test/server_side/* in this case) for server operations.

File: website/src/controllers/payments/paypal.js [hunk 1]
```diff
@@ -26,9 +26,10 @@ paypal.configure({
 });
 
 var parseErr = function(res, err){
-  //var error = err.response ? err.response.message || err.response.details[0].issue : err;
-  var error = JSON.stringify(err);
-  return res.json(400,{err:error});
+  var error = err.response ? err.response.error_description || err.response.error : 'Paypal returned an invalid error message.';
+  return res.json(400, {err: 'Paypal error: ' + error});
```
Comment (Alys) [inline website/src/controllers/payments/paypal.js]: Instead of "Paypal returned an invalid error message" I think it would be better to say just "PayPal reported an error". It's a simpler message and therefore might be less stressful to any users who see it. Also that and the "PayPal error" string should be translatable. Ask if you're not sure how to achieve that!

Please also add this string after the error message:
"Please email admin@habitica.com for assistance and paste that error message into your email."

Comment (ojeytonwilliams) [inline website/src/controllers/payments/paypal.js]: Yeah, that was a bit awkward.  The idea was that that message was there in case Paypal responded in a weird way (didn't give an error or a normal response), but I agree that it looks a bit scary.

I'll implement those changes as soon as I've wrapped my brain around the test framework!

**Comment by ojeytonwilliams** [comment]
One thing I've discovered is that the tests in sub directories of test/server_side are not being run, and when you run them a fair number of them fail.

Does anyone know if the tests are obsolete and should be deleted or are in dire need of fixing?

**Comment by crookedneighbor** [comment]
Sorry, forgot to mention that the tests in the server_side folder should not be modelled after. They were my (unsuccessful) attempt at starting unit testing for the controllers and libs.

The server code is primarily tested by integration tests, but this kind of thing will be hard to test with that.

Just do client side testing for now and I'll write something for the server before merging it in.

File: website/src/controllers/payments/paypalBillingSetup.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require('babel-core/register');
```
Comment (crookedneighbor) [inline website/src/controllers/payments/paypalBillingSetup.js]: Why did you add this here?

Comment (ojeytonwilliams) [inline website/src/controllers/payments/paypalBillingSetup.js]: That I added because I needed to run that on the command line to generate the Paypal keys and node didn't like the ES6 code.

I'm not 100% sure that's the best way to go about it, but it seemed to work.

Comment (paglias) [inline website/src/controllers/payments/paypalBillingSetup.js]: You should not use babel-core/register. When exactly this code will have to run from the command line? If only locally you can use https://babeljs.io/docs/usage/cli/

Comment (ojeytonwilliams) [inline website/src/controllers/payments/paypalBillingSetup.js]: Could you explain why babel-core/register shouldn't be used?  As I understand it, it's discouraged for production code because of how slow it is, but this particular script should never appear in production (if it does, that seems problematic).

Anyways, if you don't want that line, I'll take it out and replace it with some notes that might help out others that run into the same issues I did.

**Comment by ojeytonwilliams** [comment]
Oh, I deleted the wrong file.  Please ignore this while I sort that out!

**Comment by ojeytonwilliams** [comment]
Well, that was surprisingly frustrating.  I ended up having to force push this commit, since it refused to accept any changes.

Sorry about that, I'll try to keep it a bit cleaner in future!

File: test/spec/services/paymentServicesSpec.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      $window = {href: '', alert: sandbox.spy(), location: {search:   '', pathname: ''}};
+      $provide.value('$window', $window);
+    });
+    
+    inject(function($injector) {
+    // Set up the mock http service responses
+    $httpBackend = $injector.get('$httpBackend');
+    Payments = $injector.get('Payments');
+    
+   // console.log("PAYMENT SERVICES")
```
Comment (crookedneighbor) [inline test/spec/services/paymentServicesSpec.js:23]: What's all this for?

Comment (ojeytonwilliams) [inline test/spec/services/paymentServicesSpec.js:23]: Let's see.  

The regex is just to catch any PayPal sub requests.

beforeEach(module('habitrpg')) I copied from userServicesSpec and may not be necessary.  I didn't think to check, sorry!

The next part stubs out $window, so we can see if Angular triggers an alert when things go wrong or redirects the user when they go right.

Finally, the injector gives us access to a mock $http service and the Payments service we want to test.

Comment (crookedneighbor) [inline test/spec/services/paymentServicesSpec.js:23]: I meant, "why do you have all this commented out code"?

**Comment by crookedneighbor** [comment]
Sorry for the delay. Now that v3 of the API is out, we can finally devote some time to reviewing these PRs. Are you able to sort out the merge conflicts?

**Comment by crookedneighbor** [comment]
Going to close this, but please re-open it with the merge conflicts sorted out so we can review it.

**Comment by crookedneighbor** [comment]
Adapting this for v3. You'll receive a contributor credit if it gets merged in.

**Comment by ojeytonwilliams** [comment]
Hi.  I'll try to have a look at this as soon as I can, but I'm currently travelling.

If I can convince my windows machine to install Habitica I might be able to do something tomorrow, but sometime next week is more realistic.
